mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,110 +19,446 @@
     local src_branch="${2}"
-    local kubernetes_remote="${3:-https://github.com/kubernetes/kubernetes.git}"
+    local dst_branch="${3}"
+    local kubernetes_remote="${4:-https://github.com/kubernetes/kubernetes.git}"
+    local deps="${5:-""}"
+    local is_library="${6}"
     readonly filter src_branch
 
-    local currBranch=$(git rev-parse --abbrev-ref HEAD)
-    local previousKubeSHA=$(cat kubernetes-sha)
     local new_repo="false"
-    if [[ "${previousKubeSHA}" == "" ]]; then
+    if ! git rev-parse -q --verify HEAD; then
+        echo "Found new repo without ${dst_branch} branch, creating initial commit."
+        git-with-user commit -m "Initial commit" --allow-empty
         new_repo="true"
-        echo "sync_repo() is processing a repo that doesn't have kubernetes-sha, treat it as a new repo"
-    fi
-    
-    git remote add upstream-kube "${kubernetes_remote}" || true
-    git fetch upstream-kube
-    git branch -D kube-sync || true
-    git checkout upstream-kube/"${src_branch}" -b kube-sync
-    git reset --hard upstream-kube/"${src_branch}"
-    
-    # this command rewrites git history to *only* include $subdirectory 
-    git filter-branch -f --msg-filter 'awk 1 && echo && echo "Kubernetes-commit: ${GIT_COMMIT}"' \
-        --subdirectory-filter "${subdirectory}" HEAD
-
-    local newKubeSHA=$(git log kube-sync -1 | tail -n 1 | sed "s/Kubernetes-commit: //g")
-
-    local previousBranchSHA=$(git log --grep "Kubernetes-commit: ${previousKubeSHA}" --format='%H')
-    local commits=$(git log --no-merges --format='%H' --reverse ${previousBranchSHA}..HEAD)
-    if [ "${new_repo}" = "true" ]; then
-        commits=$(git log --no-merges --format='%H' --reverse HEAD)
-    fi
-
-    # check if any commit of commits change Godeps/Godeps.json
-    local commits_change_godeps=$(git log --format='%H' --follow Godeps/Godeps.json)
-    local cherrypicks_change_godeps="false"
-    while read commit; do
-        if [[ -z "${commit}" ]]; then
-            continue
+    else
+        echo "Starting at existing ${dst_branch} commit $(git rev-parse HEAD)."
+    fi
+
+    # remove the old kubernetes-sha file, but preserve content
+    local k_base_commit=""
+    if [ -f kubernetes-sha ]; then
+        k_base_commit=$(cat kubernetes-sha)
+        git rm -q kubernetes-sha
+        git-with-user commit -q -m "sync: remove kubernetes-sha"
+    fi
+
+    # fetch upstream kube and checkout $src_branch, name it filtered-branch
+    git remote rm upstream-kube >/dev/null || true
+    git remote add upstream-kube "${kubernetes_remote}" >/dev/null
+    git fetch -q upstream-kube -p
+    git branch -D filtered-branch >/dev/null || true
+    git branch -f upstream-branch upstream-kube/"${src_branch}"
+    echo "Checked out k8s.io/kubernetes commit $(git rev-parse upstream-branch)."
+    git checkout -q upstream-branch -b filtered-branch
+    git reset -q --hard upstream-branch
+
+    # create filtered-branch-base before filtering for
```

Comment:
It makes it easy to derive the branch point of a new branch. In kube it's easy to get that using `git merge-base`. But that commit will probably collapse onto one of its parents after filtering. To avoid having to iterate the parent until we find a commit existing after filtering, we mark the commit before filtering and let `git filter-branch` do the mapping.